core/src/main/java/io/grpc/transport/ServerTransportListener.java

```diff
@@ -48,4 +49,9 @@
    */
   ServerStreamListener streamCreated(ServerStream stream, String method,
       Metadata.Headers headers);
+
+  /**
+   * The transport completed shutting down. All resources have been released.
+   */
+  void transportTerminated();
```

Comment:
If you have methods ```isShuttingDown()``` and ```isTerminated()``` that are each thread-safe, doing ```isShuttingDown() || isTerminated()``` would yield expected results because of our knowledge of the state transitions. But ```isTerminated() || isShuttingDown()``` can race because when ```isTerminated()``` was called the state could be SHUTTING_DOWN and then when ```isShuttingDown()``` is called the state could have advanced to TERMINATED.